File: bench/suite/src/st/binary-trees/python/binary-trees.py
# binary-trees: single-threaded, pure CPython (Benchmarks Game algorithm:
# allocate full trees of nodes, walk them to count nodes).
import sys


class Node:
    __slots__ = ("left", "right")

    def __init__(self, left, right):
        self.left = left
        self.right = right


def bottom_up(depth):
    if depth > 0:
        return Node(bottom_up(depth - 1), bottom_up(depth - 1))
    return Node(None, None)


def check(t):
    if t.left is None:
        return 1
    return 1 + check(t.left) + check(t.right)


def main():
    max_depth = max(6, int(sys.argv[1]) if len(sys.argv) > 1 else 10)
    stretch = max_depth + 1
    out = ["stretch tree of depth %d\t check: %d" % (stretch, check(bottom_up(stretch)))]
    long_lived = bottom_up(max_depth)
    for d in range(4, max_depth + 1, 2):
        iterations = 1 << (max_depth - d + 4)
        c = 0
        for _ in range(iterations):
            c += check(bottom_up(d))
        out.append("%d\t trees of depth %d\t check: %d" % (iterations, d, c))
    out.append("long lived tree of depth %d\t check: %d" % (max_depth, check(long_lived)))
    print("\n".join(out))


main()

File: bench/suite/src/st/binary-trees/javascript/binary-trees.js
// binary-trees: single-threaded, plain JavaScript (Benchmarks Game algorithm:
// allocate full trees of nodes, walk them to count nodes).
'use strict';
function Node(left, right) { this.left = left; this.right = right; }
function bottomUp(depth) {
  return depth > 0 ? new Node(bottomUp(depth - 1), bottomUp(depth - 1)) : new Node(null, null);
}
function check(t) { return t.left === null ? 1 : 1 + check(t.left) + check(t.right); }

const maxDepth = Math.max(6, +process.argv[2] || 10);
const out = [];
const stretchDepth = maxDepth + 1;
out.push('stretch tree of depth ' + stretchDepth + '\t check: ' + check(bottomUp(stretchDepth)));
const longLived = bottomUp(maxDepth);
for (let d = 4; d <= maxDepth; d += 2) {
  const iterations = 1 << (maxDepth - d + 4);
  let c = 0;
  for (let i = 0; i < iterations; i++) c += check(bottomUp(d));
  out.push(iterations + '\t trees of depth ' + d + '\t check: ' + c);
}
out.push('long lived tree of depth ' + maxDepth + '\t check: ' + check(longLived));
console.log(out.join('\n'));
